File: 230-kth-smallest-element-in-a-bst/solution.cpp
#include <deque>

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  int kthSmallest(TreeNode *root, int k)
  {
    int num_visited = 0;
    std::deque<const TreeNode *> work_deque;
    const TreeNode *curr = root;
    while (curr != nullptr || !work_deque.empty())
    {
      if (curr != nullptr)
      {
        work_deque.emplace_back(curr);
        curr = curr->left;
        continue;
      }

      curr = work_deque.back();
      work_deque.pop_back();
      if (++num_visited == k)
      {
        return curr->val;
      }

      curr = curr->right;
    }

    return k;
  }
};

File: 230-kth-smallest-element-in-a-bst/solution.py
from typing import Optional
from collections import deque


class TreeNode:
  def __init__(self, val=0, left: Optional['TreeNode'] = None, right: Optional['TreeNode'] = None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
    num_visited = 0
    work_deque: deque[TreeNode] = deque()
    curr = root
    while curr is not None or work_deque:
      if curr is not None:
        work_deque.append(curr)
        curr = curr.left
        continue

      curr = work_deque.pop()
      num_visited += 1
      if num_visited == k:
        return curr.val

      curr = curr.right
